### PR TITLE
[Feature:System] Student Override Back Into Course

### PR DESCRIPTION
**Pull request by NiaHeermance**
### What is the current behavior?
When student's are dropped from a course by automated scripts, they have no way of rejoining the course without contacting the instructor. This can cause them to get behind on work, and the instructor has to manage readding students.

### What is the new behavior?
The No Access page for a course now has an option for students to rejoin a course, provided they meet certain conditions:

- The student was not manually removed from the course by the instructor
- The student was in the course to begin with
- It is within two weeks of the start of the term OR the student has accessed a gradeable, poll, forum post, course material, etc within the last 72 hours.

### Relationship to Other PRs and Issues
This PR address issue #9738 but does not close it, as future PRs will address other parts of that issue. This PR is followed by PR #9799 and PR #9783, which are independent of each other and can be merged in any order.  This PR, PR #9768, must be merged first.

### How to test?
As an instructor, move a student to the Null section and make sure you untick "Manual." Then, as that student, go to the course's page. On the No Access page you get, click the join button. Make sure that before you dropped the student, the student had done some sort of activity in the course within the past three days.

### Images
image
and then,
image

### CODE REVIEW

File: site/app/controllers/SelfRejoinController.php [hunk 1]
```diff
@@ -67,14 +67,13 @@ private function canRejoinCourse() {
...
      */
-    public function rejoinCourse(): JsonResponse {
+    public function rejoinCourse(): RedirectResponse {
+        $course_url = $this->core->buildCourseUrl();
+
         if (!$this->canRejoinCourse()) {
-            return JsonResponse::getFailResponse(
-                "This student does not meet the conditions to rejoin."
-            );
+            return new RedirectResponse("$course_url/no_access");
```
Comment (cjreed121) [inline site/app/controllers/SelfRejoinController.php:76]: It's better to use buildCourseUrl here

File: site/app/controllers/SelfRejoinController.php [hunk 2]
```diff
@@ -67,14 +67,13 @@ private function canRejoinCourse() {
...
+     * @return RedirectResponse Course url if the student met the conditions to be readded.
      */
-    public function rejoinCourse(): JsonResponse {
+    public function rejoinCourse(): RedirectResponse {
+        $course_url = $this->core->buildCourseUrl();
+
         if (!$this->canRejoinCourse()) {
-            return JsonResponse::getFailResponse(
-                "This student does not meet the conditions to rejoin."
-            );
```
Comment (cjreed121) [inline site/app/controllers/SelfRejoinController.php]: You can do core->addErrorMessage() to give the client some idea of what happened.

**Review by cjreed121 — CHANGES_REQUESTED** [review]
2 changes then this looks good to be merged

**Review by cjreed121 — APPROVED** [review]
Looks good to me. My only comment would be to add tests to ensure functionality but that could wait.

**Comment by pbailie** [comment]
This PR should have no impact on or conflict with my sysadmin scripts.

**Pull request merged**